R/zchunk_LA154.transportation_UCD.R

```diff
@@ -40,116 +44,376 @@ module_energy_LA154.transportation_UCD_DISABLED <- function(command, ...) {
+      mutate(speed.x = if_else(is.na(speed.x), speed.y, speed.x)) %>%
+      replace_na(list(speed.x = 1))
+
+    # Calculate the weighted volumes. No need to convert units because these will be aggregated and divided by weights
+    ALL_region_var <- ALL_ctry_var %>%
+      mutate(Tvkm = weight_EJ / intensity) %>%
+      mutate(Tpkm = Tvkm * `load factor`) %>%
+      mutate(Tusd = Tvkm * `non-fuel costs`) %>%
+      mutate(Thr = Tvkm / speed.x ) %>%
+      group_by(UCD_technology,UCD_fuel, UCD_sector, mode, size.class.x, year, GCAM_region_ID) %>%
+      summarise(weight_EJ = sum(weight_EJ), Tvkm = sum(Tvkm), Tpkm = sum(Tpkm),Tusd = sum(Tusd), Thr = sum(Thr))
+
+    # Reverse the calculations to calculate the weighted average of each derived variable
+    ALL_region_var <- ALL_region_var %>%
+      mutate(intensity_MJvkm = weight_EJ / Tvkm,
+             loadfactor = Tpkm / Tvkm,
+             cost_usdvkm = Tusd / Tvkm,
+             speed_kmhr = Tvkm / Thr) %>%
+      gather(variable, value, 8:16) %>%
+      select(GCAM_region_ID, UCD_sector, mode, size.class = size.class.x, UCD_technology, UCD_fuel, variable, year, value)
+
+    # Build the final data frames by variable
+    out_var_df <- split(ALL_region_var, ALL_region_var$variable) %>%
+      lapply(function(df) {select(df, -variable)})
+
+    # Part 3: Downscaling of non-motorized transport to the country level, using population
+    # Filtering population to UCD energy year and then aggregating by UCD region
+    Pop_thous_UCD_R <- L100.Pop_thous_ctry_Yh %>%
+      filter(year == energy.UCD_EN_YEAR) %>%
+      left_join_error_no_match(UCD_ctry, by = "iso") %>%
+      group_by(UCD_region) %>%
+      summarise(population = sum(value))
+
+    # Calculating non-motorized passenger-km per person
+    PKM_percap_nonmotor_UCD_R <- UCD_trn_data %>%
+      filter(mode %in% c( "Walk", "Cycle" ), variable == "service output") %>%
+      left_join_error_no_match(Pop_thous_UCD_R, by = "UCD_region") %>%
+      mutate(pkm_percap = value * CONV_BIL_MIL * CONV_MIL_THOUS / population) %>%
+      select(UCD_region, mode, year, pkm_percap)
+
+    # Compute the nonmotorized service output at the country level, using the historical population
+    PKM_nonmotor_ctry <- L100.Pop_thous_ctry_Yh %>%
+      repeat_add_columns(tibble(mode = c( "Walk", "Cycle" ))) %>%
+      left_join_error_no_match(UCD_ctry %>% select(-country_name), by = "iso")  %>%
+      left_join_error_no_match(PKM_percap_nonmotor_UCD_R %>% filter(year == energy.UCD_EN_YEAR),
+                               by = c("UCD_region", "mode")) %>%
+      mutate(value = value * 1 / CONV_MIL_THOUS * pkm_percap)
+
+    # Aggregate by GCAM region and write it out
+    PKM_nonmotor_GCAM_R <- PKM_nonmotor_ctry %>%
+      left_join_error_no_match(iso_GCAM_regID, by = "iso") %>%
+      group_by(GCAM_region_ID, mode, year = year.x) %>%
+      summarise(value = sum(value)) %>%
+      filter(year %in% HISTORICAL_YEARS)
```

Comment:
You're leaving this dataset grouped, which may (actually, does!) cause downstream problems. Please `ungroup` at end of pipeline.